test/test_auth.py

```diff
@@ -201,3 +202,216 @@ def test_scram_both(bouncer):
 
     # SCRAM password in userlist.txt
     bouncer.test(dbname="p62", user="scramuser1", password="foo")
+
+
+@pytest.mark.skipif("WINDOWS", reason="Windows does not have SIGHUP")
+def test_auth_dbname_usage_with_target_db(
+    bouncer,
+):
+    """Regression for https://github.com/pgbouncer/pgbouncer/issues/314
+
+    Check that we handle correctly attempts to use pgbouncer DB as an auth_dbname.
+    Conditions:
+        * auth_query is set
+        * stats_users is set and does not exist in the userlist.txt
+        * client connects to pgbouncer (admin DB)
+        * global auth_dbname is not set
+        * target client DB is using as an auth_dbname
+    """
+
+    config = f"""[databases]
+        [pgbouncer]
+        auth_query = SELECT usename, passwd FROM pg_shadow where usename = $1
+        auth_user = pswcheck
+        stats_users = stats
+        listen_addr = {bouncer.host}
+        admin_users = pswcheck
+        auth_type = md5
+        auth_file = {bouncer.auth_path}
+        listen_port = {bouncer.port}
+        logfile = {bouncer.log_path}
+    """
+
+    # good password
+    bouncer.test(user="pgbouncer", password="fake")
+
+    with bouncer.log_contains(
+        'cannot use the reserved "pgbouncer" database as an auth_dbname', 1
+    ):
+        with bouncer.run_with_config(config):
+            with pytest.raises(psycopg.OperationalError, match="bouncer config error"):
+                # bad password
+                # We expect that stats user does not exist in userlist.txt
+                bouncer.test(user="stats", password="stats", dbname="pgbouncer")
+
+
+@pytest.mark.skipif("WINDOWS", reason="Windows does not have SIGHUP")
+def test_auth_dbname_usage_using_autodb(
+    bouncer,
+):
+    """Regression for https://github.com/pgbouncer/pgbouncer/issues/314
```

Comment:
some tests referring to this issue do not perform any login with a stats_user. I would suggest removing this remark from those to avoid confusion in the future.

Reply:
Done